src/headers/frame_header.rs

```diff
@@ -0,0 +1,494 @@
+    epf_pass1_zeroflush: f32,
+
+    #[default(0.6)]
+    #[condition(epf_weight_custom)]
+    epf_pass2_zeroflush: f32,
+
+    #[default(false)]
+    #[condition(epf_iters > 0)]
+    epf_sigma_custom: bool,
+
+    #[default(0.46)]
+    #[condition(epf_sigma_custom && nonserialized.encoding == Encoding::VarDCT)]
+    epf_quant_mul: f32,
+
+    #[default(0.9)]
+    #[condition(epf_sigma_custom)]
+    epf_pass0_sigma_scale: f32,
+
+    #[default(6.5)]
+    #[condition(epf_sigma_custom)]
+    epf_pass2_sigma_scale: f32,
+
+    #[default(2.0 / 3.0)]
+    #[condition(epf_sigma_custom)]
+    epf_border_sad_mul: f32,
+
+    #[default(1.0)]
+    #[condition(epf_iters > 0 && nonserialized.encoding == Encoding::Modular)]
+    epf_sigma_for_modular: f32,
+
+    #[default(Extensions::default())]
+    extensions: Extensions,
+}
+
+pub struct FrameHeaderNonserialized {
+    pub xyb_encoded: bool,
+    pub num_extra_channels: u32,
+    pub have_animation: bool,
+    pub have_timecode: bool,
+    pub img_width: u32,
+    pub img_height: u32,
+}
+
+#[derive(UnconditionalCoder, Debug, PartialEq)]
+#[nonserialized(FrameHeaderNonserialized)]
+#[aligned]
+pub struct FrameHeader {
+    #[all_default]
+    #[default(true)]
+    all_default: bool,
+
+    #[default(FrameType::RegularFrame)]
+    frame_type: FrameType,
+
+    #[coder(Bits(1))]
+    #[default(Encoding::VarDCT)]
+    encoding: Encoding,
+
+    // TODO(Tomáš Král): nicer abstraction for bitflags
+    #[default(0)]
+    flags: u64,
+
+    #[default(false)]
+    #[condition(!nonserialized.xyb_encoded)]
+    do_ycbcr: bool,
+
+    #[coder(Bits(2))]
+    #[default([0, 0, 0])]
+    #[condition(do_ycbcr && flags & 0x20 == 0)]
+    jpeg_upsampling: [u32; 3],
+
+    #[coder(u2S(1, 2, 4, 8))]
+    #[default(1)]
+    #[condition(flags & 0x20 == 0)]
+    upsampling: u32,
+
+    #[size_coder(explicit(nonserialized.num_extra_channels))]
+    #[coder(u2S(1, 2, 4, 8))]
+    #[default(vec![1; nonserialized.num_extra_channels as usize])]
+    #[condition(flags & 0x20 == 0)]
+    ec_upsampling: Vec<u32>,
+
+    #[coder(Bits(2))]
+    #[default(1)]
+    #[condition(encoding == Encoding::Modular)]
+    group_size_shift: u32,
+
+    #[coder(Bits(3))]
+    #[default(3)]
+    #[condition(encoding == Encoding::VarDCT && nonserialized.xyb_encoded)]
+    x_qm_scale: u32,
+
+    #[coder(Bits(3))]
+    #[default(2)]
+    #[condition(encoding == Encoding::VarDCT && nonserialized.xyb_encoded)]
+    b_qm_scale: u32,
+
+    #[condition(frame_type != FrameType::ReferenceOnly)]
+    #[default(Passes::default())]
+    passes: Passes,
+
+    #[coder(u2S(1, 2, 3, 4))]
+    #[default(0)]
+    #[condition(frame_type == FrameType::LFFrame)]
+    lf_level: u32,
+
+    #[default(false)]
+    #[condition(frame_type != FrameType::LFFrame)]
+    have_crop: bool,
+
+    #[coder(u2S(Bits(8), Bits(11) + 256, Bits(14) + 2304, Bits(30) + 18688))]
+    #[default(0)]
+    #[condition(have_crop && frame_type != FrameType::ReferenceOnly)]
+    x0: u32,
```

Comment:
I *think* this should be `i32` (it can be negative, for sure -- although IIRC in libjxl we never generate such files)

Reply:
Hmm, it's a bit surprising, but x0 and y0 are `int32_t` in libjxl.

Reply:
Yes, that can be useful to represent i.e. layers in a saved file of something like GIMP :)